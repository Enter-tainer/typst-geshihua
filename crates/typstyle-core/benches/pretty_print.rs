use criterion::{criterion_group, criterion_main, Criterion};
use std::fs;
use std::time::Duration;
use typst_syntax::Source;
use typstyle_core::{attr::AttrStore, Typstyle};

fn bench_attrs(c: &mut Criterion, id: &str, path: &str) {
    c.bench_function(id, |b| {
        let content = fs::read_to_string(path).unwrap();
        let source = Source::detached(content);
        b.iter(|| AttrStore::new(source.root()))
    });
}

fn bench_pretty(c: &mut Criterion, id: &str, path: &str) {
    fn pretty_print_source(source: Source) -> String {
        let t = Typstyle::new_with_src(source, 80);
        t.pretty_print()
    }

    c.bench_function(id, |b| {
        let content = fs::read_to_string(path).unwrap();
        let source = Source::detached(content);
        b.iter(|| pretty_print_source(source.clone()))
    });
}

const TEST_ASSETS: [&str; 8] = [
    "tablex",
    "cetz-manual",
    "undergraduate-math",
    "packages/codly",
    "packages/fletcher-diagram",
    "packages/fletcher-draw",
    "packages/touying/core",
    "packages/touying/utils",
];

fn benchmark_attrs(c: &mut Criterion) {
    for name in TEST_ASSETS {
        bench_attrs(
            c,
            &format!("attrs-{name}"),
            &format!("../../tests/assets/{name}.typ"),
        );
    }
}

fn benchmark_pretty(c: &mut Criterion) {
    for name in TEST_ASSETS {
        bench_pretty(
            c,
            &format!("pretty-{name}"),
            &format!("../../tests/assets/{name}.typ"),
        );
    }
}

criterion_group! {
    name = benches;
    config = Criterion::default()
        .measurement_time(Duration::from_secs(8))  // 只增加测量时间
        .sample_size(50);                          // 减少采样数量
    targets = benchmark_attrs, benchmark_pretty
}
criterion_main!(benches);
