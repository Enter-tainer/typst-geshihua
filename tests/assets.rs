extern crate libtest_mimic;

use libtest_mimic::{Arguments, Failed, Trial};
use typst_geshihua::pretty_print;

use std::{env, error::Error, ffi::OsStr, fs, path::Path};

fn main() -> Result<(), Box<dyn Error>> {
    let args = Arguments::from_args();
    let tests = collect_tests()?;
    libtest_mimic::run(&args, tests).exit();
}

/// Creates one test for each `.typ` file in the current directory or
/// sub-directories of the current directory.
fn collect_tests() -> Result<Vec<Trial>, Box<dyn Error>> {
    fn visit_dir(path: &Path, tests: &mut Vec<Trial>) -> Result<(), Box<dyn Error>> {
        for entry in fs::read_dir(path)? {
            let entry = entry?;
            let file_type = entry.file_type()?;

            // Handle files
            let path = entry.path();
            if file_type.is_file() {
                if path.extension() == Some(OsStr::new("typ")) {
                    let name = path
                        .strip_prefix(env::current_dir()?)?
                        .display()
                        .to_string();

                    let test_40 = {
                        let path: std::path::PathBuf = path.clone();
                        Trial::test(format!("{} - 40char", name), move || check_file(&path, 40))
                            .with_kind("typst")
                    };
                    let test_80 = {
                        let path = path.clone();
                        Trial::test(format!("{} - 80char", name), move || check_file(&path, 80))
                            .with_kind("typst")
                    };
                    let test_120 = {
                        let path = path.clone();
                        Trial::test(format!("{} - 120char", name), move || {
                            check_file(&path, 120)
                        })
                        .with_kind("typst")
                    };
                    let test_convergence =
                        Trial::test(format!("{} - convergence", name), move || {
                            check_convergence(&path, 80)
                        });
                    tests.extend([test_40, test_80, test_120, test_convergence]);
                }
            } else if file_type.is_dir() {
                // Handle directories
                visit_dir(&path, tests)?;
            }
        }

        Ok(())
    }

    // We recursively look for `.typ` files, starting from the current
    // directory.
    let mut tests = Vec::new();
    let current_dir = env::current_dir()?;
    visit_dir(&current_dir, &mut tests)?;

    Ok(tests)
}

/// Performs a couple of tidy tests.
fn check_file(path: &Path, width: usize) -> Result<(), Failed> {
    let content = fs::read(path).map_err(|e| format!("Cannot read file: {e}"))?;

    // Check that the file is valid UTF-8
    let content = String::from_utf8(content)
        .map_err(|_| "The file's contents are not a valid UTF-8 string!")?;
    let rel_path = pathdiff::diff_paths(path, env::current_dir().unwrap()).unwrap();
    let doc_string = pretty_print(&content, width);
    let replaced_path = rel_path
        .to_str()
        .unwrap()
        .replace(std::path::MAIN_SEPARATOR, "-");
    insta::with_settings!({
        snapshot_suffix => format!("{}-{width}", replaced_path),
        input_file => path,
    }, {
        insta::assert_snapshot!(doc_string);
    });
    Ok(())
}

fn check_convergence(path: &Path, width: usize) -> Result<(), Failed> {
    let content = fs::read(path).map_err(|e| format!("Cannot read file: {e}"))?;

    // Check that the file is valid UTF-8
    let content = String::from_utf8(content)
        .map_err(|_| "The file's contents are not a valid UTF-8 string!")?;
    let first_pass = pretty_print(&content, width);
    let second_pass = pretty_print(&first_pass, width);
    pretty_assertions::assert_str_eq!(
        first_pass,
        second_pass,
        "first pass and second pass are not the same!"
    );
    Ok(())
}
