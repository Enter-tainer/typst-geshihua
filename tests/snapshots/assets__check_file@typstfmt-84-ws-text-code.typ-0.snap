---
source: tests/assets.rs
expression: doc_string
input_file: tests/assets/typstfmt/84-ws-text-code.typ
---
#let f() = []
a#sym.RR

a#f()
